students/ZidanLuo/session09/donor_models.py

```diff
@@ -0,0 +1,193 @@
+#!/usr/bin/env python
+import sys
+import os
+
+class Donor_Collection():
+    def __init__(self):
+        self.donor_dict = {"Fred Jones": Donor("Fred Jones", 500)}
+        self.sorted_dict = {"Fred Jones": Donor("Fred Jones", 500)}
+
+        self.longest_name = "Fred Jones"
+        self.largest_amount = 500
+
+    
+    def copy_constructor(self, sampleDB, longest_name, largest_amount):
+        """
+        A copy constructor take in sample database for testin
+
+        :param: a dictionary collection of initialized donation, longest name in sample database, 
+        largest donation amount in the sample database
+         :return: 
+        """
+        self.donor_dict = sampleDB
+        self.longest_name = longest_name
+        self.largest_amount = largest_amount
+
+    
+    def add_donor(self, donor_name, donation_amount):
+        """
+        Add a donor to the donor collection, and compare the length of the donor's name and donation 
+        amount to decide the widths we are going to use in generating reports
+
+        :param: name of the donor and donation amount
+        :return: the new donor database
+        """
+        if donor_name in self.donor_dict:
+            self.donor_dict[donor_name].add_donation(donation_amount)
+        else:
+            self.donor_dict[donor_name] = Donor(donor_name, donation_amount)
+
+        if len(donor_name) > len(self.longest_name):
```

Comment:
probably better to compute this when you are making the report -- you may add / remove many donors in between reports...